Add example of Custom Kafka Headers

oleg-rakitin opened the pull request.

Commit:
Add example of Custom Kafka Headers

akurilov commented on README.md:
````diff
@@ -78,4 +78,50 @@ Not supported
+```javascript
+var customKafkaHeadersConfig = {
+    "storage" : {
+        "net" : {
````
let's use `storage-driver-request-headers` either `storage-driver-create-headers` option
also, the doc is updated, but the option is not added into the defaults/schema yaml files
moreover, the declared configuration option should be handled by the driver's code

AndreyKoltsov1997 reviewed (dismissed):
What about the record key?

Commit:
Add headers handling, add options in yml files


"Signed-off-by: [Oleg Rakitin] [limonchik16@gmail.com]"

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Commit:
Add headers handling

oleg-rakitin committed:
Indentation correction

"Signed-off-by: [Oleg Rakitin] [limonchik16@gmail.com]" 

Developer's Certificate of Origin 1.1 

By making a contribution to this project, I certify that: 

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or 

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or 

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it. 

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

oleg-rakitin committed:
Fix a forgotten bracket

"Signed-off-by: [Oleg Rakitin] [limonchik16@gmail.com]" 

Developer's Certificate of Origin 1.1 

By making a contribution to this project, I certify that: 

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or 

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or 

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it. 

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

oleg-rakitin committed:
Update KafkaStorageDriver.java

oleg-rakitin committed:
Update KafkaStorageDriver.java

oleg-rakitin committed:
Fix errors

"Signed-off-by: [Oleg Rakitin] [limonchik16@gmail.com]"

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

oleg-rakitin committed:
Merge branch 'master' into KAFKA-32-make-the-producer-record-key-and-headers-configurable

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java:
```diff
@@ -31,6 +37,22 @@ public KafkaStorageDriver(
+              || headerValue.contains(INIT_MARKER)) {
+        dynamicHeaders.put(headerKey, headerValue);
+      } else {
+        throw new IllegalArgumentException("Unsupported key");
```
the constant headers should be used also

oleg-rakitin committed:
Add sharedHeaders

"Signed-off-by: [Oleg Rakitin] [limonchik16@gmail.com]" 

Developer's Certificate of Origin 1.1 

By making a contribution to this project, I certify that: 

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or 

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or 

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it. 

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Commit:
Fix conflicts

Commit:
Fixes

"Signed-off-by: [Oleg Rakitin] [limonchik16@gmail.com]"

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

akurilov commented on src/main/resources/config/defaults-storage-driver-kafka.yml:
```diff
@@ -7,6 +7,10 @@ storage:
       key:
         enabled: false
+      headers:
+        Connection: Keep-Alive
```
this is HTTP-specific headers which are not suitable for kafka
let's leave the headers empty

oleg-rakitin committed:
Change the headers to the empty

"Signed-off-by: [Oleg Rakitin] [limonchik16@gmail.com]"

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

oleg-rakitin committed:
Merge branch 'master' into KAFKA-32-make-the-producer-record-key-and-headers-configurable

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java:
```diff
@@ -65,6 +72,21 @@ public KafkaStorageDriver(
     super(testStepId, dataInput, storageConfig, verifyFlag, batchSize);
     var driverConfig = storageConfig.configVal("driver");
+	final var headersMap = driverConfig.<String>mapVal("headers");
+    for (final var header : headersMap.entrySet()) {
```
the default value will cause NullPointerException here

akurilov commented on src/main/resources/config/defaults-storage-driver-kafka.yml:
```diff
@@ -7,6 +7,7 @@ storage:
     create:
       key:
         enabled: false
+      headers: null
```
https://stackoverflow.com/questions/33510094/syntax-for-empty-dictionary-in-yaml

oleg-rakitin committed:
Change zlib version

"Signed-off-by: [Oleg Rakitin] [limonchik16@gmail.com]" 

Developer's Certificate of Origin 1.1 

By making a contribution to this project, I certify that: 

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or 

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or 

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it. 

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

oleg-rakitin committed:
Merge branch 'master' into KAFKA-32-make-the-producer-record-key-and-headers-configurable

oleg-rakitin committed:
Check for empty default headers

"Signed-off-by: [Oleg Rakitin] [limonchik16@gmail.com]"

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.
